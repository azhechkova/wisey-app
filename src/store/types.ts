import { TrackVideoType } from '~/types';

export type RootState = {
  app: {
    activeLesson: string;
    videoProgress: TrackVideoType[];
  };
};
