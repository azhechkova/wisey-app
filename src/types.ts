export type LessonType = {
  id: string;
  title: string;
  duration: number;
  order: number;
  type: 'video';
  status: 'unlocked' | 'locked';
  link: string;
  previewImageLink: string;
  meta: object | null;
};

export type CourseType = {
  id: string;
  title: string;
  tags: string[];
  launchDate?: string;
  status: 'launched';
  description: string;
  duration: number;
  lessonsCount?: number;
  lessons: LessonType[];
  containsLockedLessons: boolean;
  previewImageLink: string;
  rating: number;
  meta?: {
    slug?: string;
    skills?: string[];
    courseVideoPreview?: {
      link?: string;
      previewImageLink?: string;
      duration?: number;
    };
  } | null;
};

export type ErrorType = unknown | null | Error;

export type TApiResponse<T> = {
  loading: boolean;
  error: ErrorType;
  data: T | null;
};
