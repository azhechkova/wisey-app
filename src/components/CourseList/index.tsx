import { Box } from '@mui/material';
import React from 'react';

import useStyles from './styles';
import { CourseType } from '../../types';
import { Course } from '..';

interface CourseListProps {
  courses: CourseType[];
}

const CourseList = ({ courses }: CourseListProps): JSX.Element => {
  const { classes } = useStyles();

  return (
    <Box className={classes.list}>
      {courses.map(course => (
        <Course key={course.id} course={course} />
      ))}
    </Box>
  );
};

export default CourseList;
