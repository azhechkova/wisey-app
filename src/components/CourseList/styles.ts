import { makeStyles } from 'tss-react/mui';

const useStyles = makeStyles()(() => ({
  list: {
    display: 'grid',
    gridTemplateColumns: 'repeat(3, 1fr)',
    gap: '1em',
    '@media (max-width: 992px)': {
      gridTemplateColumns: 'repeat(2, 1fr)',
    },
    '@media (max-width: 572px)': {
      gridTemplateColumns: 'repeat(1, 1fr)',
    },
  },
}));

export default useStyles;
