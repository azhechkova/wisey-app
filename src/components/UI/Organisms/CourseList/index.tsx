import React, { useState, useMemo } from 'react';
import { Box, Button } from '@mui/material';

import { CourseType } from '~/types';

import Course from '~/components/UI/Molecules/Course';

import useStyles from './styles';

interface CourseListProps {
  courses: CourseType[];
}

const COURSES_PER_PAGE = 10;

const CourseList = ({ courses }: CourseListProps): JSX.Element => {
  const { classes } = useStyles();
  const [limit, setLimit] = useState(COURSES_PER_PAGE);

  const onShowMore = (): void => {
    setLimit(prev => prev + COURSES_PER_PAGE);
  };

  const filteredList = useMemo(
    () => (courses ? courses.slice(0, limit) : []),
    [limit, courses],
  );

  return (
    <Box>
      <Box className={classes.list}>
        {filteredList.map(course => (
          <Course key={course.id} course={course} />
        ))}
      </Box>
      {limit < courses?.length && (
        <Button
          className={classes.button}
          variant="outlined"
          onClick={onShowMore}
        >
          Show more
        </Button>
      )}
    </Box>
  );
};

export default CourseList;
