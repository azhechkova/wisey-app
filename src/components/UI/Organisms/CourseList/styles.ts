import { makeStyles } from 'tss-react/mui';

const useStyles = makeStyles()(() => ({
  list: {
    display: 'grid',
    gridTemplateColumns: 'repeat(3, 1fr)',
    gap: '1em',
    margin: '50px 0',
    '@media (max-width: 992px)': {
      gridTemplateColumns: 'repeat(2, 1fr)',
    },
    '@media (max-width: 572px)': {
      gridTemplateColumns: 'repeat(1, 1fr)',
    },
  },
  button: {
    margin: '0 auto',
    display: 'block',
  },
}));

export default useStyles;
