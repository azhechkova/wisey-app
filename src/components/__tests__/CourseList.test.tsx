import React from 'react';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { CourseType } from '~/types';
import CourseList from '../UI/Organisms/CourseList';

const course: CourseType = {
  id: '1',
  title: 'Test Course 1',
  description: 'This is a test course',
  previewImageLink: 'https://example.com/image1.png',
  tags: [],
  status: 'launched',
  duration: 50,
  lessons: [],
  containsLockedLessons: false,
  rating: 5,
};
const courses: CourseType[] = Array(11).map((_, index) => {
  return { ...course, id: `${index}` };
});

describe('CourseList component', () => {
  it('should render the correct number of courses', () => {
    const { container } = render(
      <MemoryRouter>
        <CourseList courses={courses} />
      </MemoryRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render the "Show more" button when there are more courses than the limit', () => {
    render(
      <MemoryRouter>
        <CourseList courses={courses} />
      </MemoryRouter>,
    );
    const button = screen.getByText('Show more');

    expect(button).toBeInTheDocument();
  });
});
