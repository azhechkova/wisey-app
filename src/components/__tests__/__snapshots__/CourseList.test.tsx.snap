// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseList component should render the correct number of courses 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiBox-root css-1r1jkbv-list"
    />
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-j6m4ra-MuiButtonBase-root-MuiButton-root-button"
      tabindex="0"
      type="button"
    >
      Show more
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
